### PR TITLE
[IMP] account_invoice_merge: refactor invoice merging

### PR DESCRIPTION
**Pull request by mpanarin**
Refactor of this module. It used deprecated methods, was obscure and hard to extend. 
Now it is much cleaner and readable, switching to SavepointCase tests helped a lot as well (from ~30 seconds to ~4)

### CODE REVIEW

**Comment by mpanarin** [comment]
linter is erroring in other modules that I didn't touch. But the last build in this repo is green(2 weeks old). Did something change in linter settings?

**Comment by MeetKD** [comment]
Bump Version number up in manifest file.

**Comment by mpanarin** [comment]
@MeetKD updated

File: account_invoice_merge/__manifest__.py [hunk 1]
```diff
@@ -8,13 +8,15 @@
 
 {
     'name': 'Account Invoice Merge',
-    'version': '10.0.1.0.0',
+    'version': '10.0.2.0.0',
     'category': 'Finance',
     'summary': "Merge invoices in draft",
     'author': "Elico Corp,Odoo Community Association (OCA)",
     'website': 'http://www.openerp.net.cn',
```
Comment (elicoidal) [inline account_invoice_merge/__manifest__.py]: should be https://github.com/OCA/account-invoicing

File: account_invoice_merge/models/__init__.py [hunk 2]
```diff
@@ -7,3 +7,4 @@
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
 from . import account_invoice
+from . import account_invoice_lines
```
Comment (elicoidal) [inline account_invoice_merge/models/__init__.py]: copyright headers no mandatory in init.py

**Review by elicoidal — APPROVED** [review]
small details. LGTM

File: account_invoice_merge/__manifest__.py [hunk 3]
```diff
@@ -8,13 +8,15 @@
...
+    'version': '10.0.2.0.0',
     'category': 'Finance',
     'summary': "Merge invoices in draft",
     'author': "Elico Corp,Odoo Community Association (OCA)",
-    'website': 'http://www.openerp.net.cn',
+    'website': 'https://github.com/OCA/account-invoicing',
     'license': 'AGPL-3',
-    'depends': ['account'],
+    'depends': [
+        'account'
```
Comment (ghost) [inline account_invoice_merge/__manifest__.py]: comma is missing

File: account_invoice_merge/models/account_invoice_lines.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            'discount',
+            'invoice_line_tax_ids',
+            'price_unit',
+            'product_id',
+            'account_id',
+            'account_analytic_id',
+            'uom_id',
+            'analytic_tag_ids',
+        ]
+        for field in ['sale_line_ids']:
```
Comment (ghost) [inline account_invoice_merge/models/account_invoice_lines.py]: what's the reason to loop over a single item list? :smile:

File: account_invoice_merge/models/account_invoice_lines.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline account_invoice_merge/models/account_invoice_lines.py]: don't u want to rename the file after a model, by chance?

File: account_invoice_merge/models/account_invoice.py [hunk 6]
```diff
@@ -4,214 +4,180 @@
...
+                ),
+            })
+            lines = []
+            # merge invoice lines
+            line_ids = group.mapped("invoice_line_ids")
+            for line_group in line_ids.group_by_fields():
+                line_vals = line_group[0].get_group_values()
+                line_vals['quantity'] = sum(line_group.mapped('quantity'))
+                if remove_empty_invoice_lines:
+                    # TODO: not sure this check is needed as
```
Comment (leemannd) [inline account_invoice_merge/models/account_invoice.py:162]: Did you check it?
If you create an invoice with invoice lines, can you later modify them with a `0` quantity?

**Review by leemannd — APPROVED** [review]
LGTM, there is only the lint to fix

**Comment by gurneyalex** [comment]
@leemannd @mpanarin merging #389 should fix your PR.

**Comment by pedrobaeza** [comment]
Please rebase

**Comment by mpanarin** [comment]
@pedrobaeza done

**Comment by pedrobaeza** [comment]
I avoid this module as hell since v9. What is your use case for using it having the new invoicing system from SO/PO?

**Review by TDu — APPROVED** [review]
Great job :+1:

**Comment by fclementic2c** [comment]
@pedrobaeza we want to merge invoices generated by subscriptions. 
In our business case, we can have several subscriptions for 1 customer but we want to send only 1 grouped invoice.

...saying that I do not like very much this feature too but I do not see any other options :\

**Comment by pedrobaeza** [comment]
Yeah, I see. I'm supposing you are using enterprise subscription module. FYI, we have this solved in OCA module contract at https://github.com/OCA/contract, used for merging together at invoice generation recurring lines + sales orders pending to invoice. Another option is to add the feature of having different recurring interval at line level of the contract. A draft is available here: https://github.com/OCA/contract/pull/109

**Comment by fclementic2c** [comment]
@pedrobaeza Interesting. thanks

cc @gurneyalex

**Comment by pedrobaeza** [comment]
Reference of the "invoice merging" inside contract (much more simple than this):

https://github.com/OCA/contract/blob/6a6192ec414dc1a93a1a13248261c7d5187fe3cf/contract/models/account_analytic_account.py#L258-L262

**Comment by MiquelRForgeFlow** [comment]
What is the status of this PR? Will it get merged?

File: account_invoice_merge/models/account_invoice_line.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            'name',
+            'origin',
+            'discount',
+            'invoice_line_tax_ids',
+            'price_unit',
+            'product_id',
+            'account_id',
+            'account_analytic_id',
+            'uom_id',
+            'analytic_tag_ids',
```
Comment (intero-chz) [inline account_invoice_merge/models/account_invoice_line.py:37]: Why not copying `sequence` too?

**Comment by intero-chz** [comment]
Very good refactor, will it be merged?

**Comment by elicoidal** [comment]
There is a conflict to solve but no reason not to merge it AFAIK

File: account_invoice_merge/models/account_invoice_line.py (hunk 7 above)
Comment (elicoidal) [reply account_invoice_merge/models/account_invoice_line.py:37]: Does it really make sense in the new invoice the line sequence from all other invoices?

Comment (intero-chz) [reply account_invoice_merge/models/account_invoice_line.py:37]: The more i think about it - no, it depends on the customer using it. I have a customer where this is needed, but with this refactor more parts can be manipulated or extended. That closure in the original code is a pain in the ass.

Comment (elicoidal) [reply account_invoice_merge/models/account_invoice_line.py:37]: @intero-chz @mpanarin can I merge without this?

Comment (intero-chz) [reply account_invoice_merge/models/account_invoice_line.py:37]: @elicoidal for me this PR is fine, but i can't do something about the merge itself :baby:

**Comment by elicoidal** [comment]
OK. Let's move forward as it has been stalled for too long

**Comment by rousseldenis** [comment]
@elicoidal @mpanarin The 10.0 branch is red on travis as there was an error in account_invoice_merge_purchase module tests.
Is someone could fix it as this leads to red PR's... Thanks!

**Comment by intero-chz** [comment]
@elicoidal maybe a revert would help. Besides the *merge_attachment module, the other two modules should be refactored very easy. Would you like to do that, or should i take a look into that?

**Comment by elicoidal** [comment]
I can revert. For the refactor, I cannot do it (not a dev myself :wink: )

**Comment by intero-chz** [comment]
yes i see @mpanarin was making the refactor. I'll see what i can do on sunday. I have another little thing for mass_editing in my pipeline.

**Pull request merged**